statistics/row_sampler.go

```diff
@@ -0,0 +1,306 @@
+	collector := &RowSampleCollector{
+		Samples:       make(WeightedRowSampleHeap, 0, s.MaxSampleSize),
+		NullCount:     make([]int64, len(s.ColsFieldType)+len(s.ColGroups)),
+		FMSketches:    make([]*FMSketch, 0, len(s.ColsFieldType)+len(s.ColGroups)),
+		TotalSizes:    make([]int64, len(s.ColsFieldType)+len(s.ColGroups)),
+		MaxSampleSize: s.MaxSampleSize,
+	}
+	for i := 0; i < len(s.ColsFieldType)+len(s.ColGroups); i++ {
+		collector.FMSketches = append(collector.FMSketches, NewFMSketch(s.MaxFMSketchSize))
+	}
+	ctx := context.TODO()
+	chk := s.RecordSet.NewChunk()
+	it := chunk.NewIterator4Chunk(chk)
+	for {
+		err := s.RecordSet.Next(ctx, chk)
+		if err != nil {
+			return nil, err
+		}
+		if chk.NumRows() == 0 {
+			return collector, nil
+		}
+		collector.Count += int64(chk.NumRows())
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			datums := RowToDatums(row, s.RecordSet.Fields())
+			for i, val := range datums {
+				// For string values, we use the collation key instead of the original value.
+				if s.Collators[i] != nil && !val.IsNull() {
+					decodedVal, err := tablecodec.DecodeColumnValue(val.GetBytes(), s.ColsFieldType[i], s.Sc.TimeZone)
+					if err != nil {
+						return nil, err
+					}
+					decodedVal.SetBytesAsString(s.Collators[i].Key(decodedVal.GetString()), decodedVal.Collation(), uint32(decodedVal.Length()))
+					encodedKey, err := tablecodec.EncodeValue(s.Sc, nil, decodedVal)
+					if err != nil {
+						return nil, err
+					}
+					val.SetBytes(encodedKey)
+				}
+			}
+			err := collector.collectColumns(s.Sc, datums)
+			if err != nil {
+				return nil, err
+			}
+			err = collector.collectColumnGroups(s.Sc, datums, s.ColGroups)
+			if err != nil {
+				return nil, err
+			}
+			weight := s.Rng.Int63()
+			collector.sampleRow(datums, weight)
+		}
+	}
+}
+
+func (s *RowSampleCollector) collectColumns(sc *stmtctx.StatementContext, cols []types.Datum) error {
+	for i, col := range cols {
+		if col.IsNull() {
+			s.NullCount[i]++
+		}
+		s.TotalSizes[i] += int64(len(col.GetBytes())) - 1
+		// Minus one is to remove the flag byte.
+		err := s.FMSketches[i].InsertValue(sc, col)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *RowSampleCollector) collectColumnGroups(sc *stmtctx.StatementContext, cols []types.Datum, colGroups [][]int64) error {
+	colLen := len(cols)
+	datumBuffer := make([]types.Datum, 0, len(cols))
+	for i, group := range colGroups {
+		datumBuffer = datumBuffer[:0]
+		allNull := true
+		for _, c := range group {
+			datumBuffer = append(datumBuffer, cols[c])
+			allNull = allNull && cols[c].IsNull()
+			s.TotalSizes[colLen+i] += int64(len(cols[c].GetBytes())) - 1
+		}
+		err := s.FMSketches[colLen+i].InsertRowValue(sc, datumBuffer)
+		if err != nil {
+			return err
+		}
+		if allNull {
+			s.NullCount[colLen+i]++
+		}
+	}
+	return nil
+}
+
+func (s *RowSampleCollector) sampleRow(cols []types.Datum, weight int64) {
```

Comment:
Can we implement this function by calling `sampleZippedRow`? the core TopK logic is shared by these 2 functions.

Reply:
Actually, we could. The `sampleRow` is only used by unistore. We don't need to consider too much about its performance.